src/counter.c

```diff
@@ -0,0 +1,98 @@
+// Commands for counting edges on GPIO input pins
+//
+// Copyright (C) 2021  Adrian Keet <arkeet@gmail.com>
+//
+// This file may be distributed under the terms of the GNU GPLv3 license.
+
+#include <stdbool.h>
+#include "basecmd.h" // oid_alloc
+#include "board/gpio.h" // struct gpio_adc
+#include "board/irq.h" // irq_disable
+#include "board/misc.h" // timer_read_time
+#include "command.h" // DECL_COMMAND
+#include "sched.h" // DECL_TASK
+
+struct counter {
+    struct timer timer;
+    struct gpio_in pin;
+    uint32_t poll_ticks;
+    uint32_t sample_ticks, last_sample_time;
+    uint32_t count, last_count_time;
+    bool value          : 1;
+    bool sample_pending : 1;
+};
+
+static struct task_wake counter_wake;
+
+static uint_fast8_t
+counter_event(struct timer *timer)
+{
+    struct counter *c = container_of(timer, struct counter, timer);
+
+    uint32_t time = c->timer.waketime;
+
+    uint8_t value = gpio_in_read(c->pin);
+    // possibly useful invariant: count % 2 == value
+    if (c->value != value) {
+        c->value = value;
+        c->count++;
+        c->last_count_time = time;
+    }
+
+    if (time - c->last_sample_time >= c->sample_ticks) {
```

Comment:
This should use `timer_is_before()`.  It's preferable to use that call instead of direct comparisons so as to avoid any issues with counter rollover.